[rplidar users] Potential errors with drivers + 360 lidar support

**Required Info:**

- Operating System:
  - Ubuntu 16.04
- Installation type:
  - from source
- ROS Version
  - ROS Kinetic
- Version or commit hash:
  - kinetic-devel, latest commit (9b14cf8369a63264fd9935be4cca5cf14386bfc5)

#### Steps to reproduce issue
I created a minimal repo with bags, config and launch files:  
https://github.com/saschroeder/turtlebot_slam_testing  

The bug always occurs with rosbags and sometimes in simulation, too. Disabling loop closure prevents the bug.   
I can upload more rosbags if necessary.

#### Expected behavior
Record an accurate map and correctly use loop closure to improve the map.

#### Actual behavior
After some time the map is rotated around 180° (may be related or not) and some time after this (seconds to minutes) the pose graph is optimized in a wrong way, causing the map to be broken (see images below).  

#### Additional information

Example 1 (before/after optimization):  
![ex1_before_optimization](https://user-images.githubusercontent.com/28823855/80203283-363c7600-8627-11ea-8c38-02f9c35430bd.png)
![ex1_after_optimization](https://user-images.githubusercontent.com/28823855/80203276-33da1c00-8627-11ea-8f56-8496ef30d2bf.png)

Example 2 (before/ after optimization):
![ex2_before_optimization](https://user-images.githubusercontent.com/28823855/80203289-3a689380-8627-11ea-8ed9-2dc2d7175e99.png)
![ex2_after_optimization](https://user-images.githubusercontent.com/28823855/80203287-39376680-8627-11ea-97ba-d8a09711c7fa.png)

Update:
I am a little confused because I feel like my results are not consistent now, but perhaps I was overlooking something before.

**It appear now to be related to high CPU load.**
Up until Tuesday, it failed consistently on both the robot and my workstation laptops. On Tuesday it would not fail on my workstation laptop.
Today I started artificially loading up the CPU on my i78850H (6 cores/12 threads, up to 4Ghz), and that does force the problem to happen.

If it was CPU related, that would explain why it is a bit random. My robot's CPU is slower, and even my powerful laptop is subject to both whatever else I'm running and thermal throttling.

I didn't think I saw this error before, but now I see it every time before the map "goes out of sync":
```
[ERROR] [1596036554.073900666, 1595664385.405057103]: Transform from base_link to odom failed: Lookup would require extrapolation into the past.  Requested time 1595664310.081902091 but the earliest data is at time 1595664355.599803672, when looking up transform from frame [base_footprint] to frame [odom]
```
and as you know this is from line https://github.com/SteveMacenski/slam_toolbox/blob/noetic-devel/slam_toolbox/src/slam_toolbox_common.cpp#L358 after which it sets an "identity frame" and now the map to odometry transform is zeroed out and nothing can proceed.

So two thoughts, assuming this is really where things go badly:
1. I presume this should run fine with the default config on modern but not insane computer?
My only two changes to the config from default are:
```
resolution: 0.025 # 0.05
minimum_travel_distance: 0.1 # 0.5
minimum_travel_heading: 0.1 # 0.5
```
I increased the resolution because it appears to help with path planning through doorways. They aren't particularly narrow, but the path planners get hung up on corners when the pixels are big.

I decreased the travel distance because the path planner seemed to get stalled when the map wasn't updating during slow movement.

Are either of these adding an exponential increase in CPU usage? If so, do you have some guidance on tuning them? Like, how much is too much?

2. Is this the best or only possible response to this failure?
I feel like wiping out the entire map to odom transform is draconian, but obviously I'm looking at one very specific situation, and I presume you understand the reasoning better.
In my very naive view of the situation, I would want to just not send out any update or repeat the previous transform until I had data I could rely on, rather than making a very real and drastic change to set the transform to all zeros.

I have **seen** the map "lose sync" before and then recover, but I have not been able to reproduce that via the bag file yet. So maybe you expect that to happen, but it isn't for some reason?


I am going to play around with some of my naive ideas, and I will update you again if I find either more data or a solution that at least patches this up for me.

Meanwhile, when you do get to this (**and I am grateful if you get to it any time in the next few months, no rush at all**), hopefully this tells you where to look, and that you will have to try it on a slow or overloaded CPU.

**Update:**
1. Reducing the resolution to 0.05 does decrease the CPU load, and hence requires me to load up the CPU a lot more before I can cause the symptom, but it is still possible and repeatable.
2. If I stop all CPU loading after the map to odom transform was zeroed out, it does sometimes recover.
3. I honestly don't know what this has to do with the RPLIDAR. Perhaps I erroneously jumped on to this issue when mine is unrelated? Unless there is a convergence of things that I'm not aware of. I presume this reset of the transform is what happens to anybody when their CPU is busy?

Also, this code is 10 years old, so I guess it is battle tested?
![image](https://user-images.githubusercontent.com/6188278/88829168-d90a1080-d191-11ea-88d4-9a98f93537a1.png)

I'm going to dig into the error a bit more, because I don't fully understand it yet. Maybe when I do I can understand the process better:
```
[ERROR] [1596041206.551335169, 1595664078.661087607]: Transform from base_link to odom failed: Lookup would require extrapolation into the past.  Requested time 1595664048.240699958 but the 
earliest data is at time 1595664048.876555413, when looking up transform from frame [base_footprint] to frame [odom]
```

--- a/slam_toolbox/lib/karto_sdk/include/karto_sdk/Karto.h
+++ b/slam_toolbox/lib/karto_sdk/include/karto_sdk/Karto.h
@@ -5634,16 +5634,7 @@ namespace karto
      */
     void SetSensorPose(const Pose2& rScanPose)
     {
-      Pose2 deviceOffsetPose2 = GetLaserRangeFinder()->GetOffsetPose();
-      kt_double offsetLength = deviceOffsetPose2.GetPosition().Length();
-      kt_double offsetHeading = deviceOffsetPose2.GetHeading();
-      kt_double angleoffset = atan2(deviceOffsetPose2.GetY(), deviceOffsetPose2.GetX());
-      kt_double correctedHeading = math::NormalizeAngle(rScanPose.GetHeading());
-      Pose2 worldSensorOffset = Pose2(offsetLength * cos(correctedHeading + angleoffset - offsetHeading),
-                                      offsetLength * sin(correctedHeading + angleoffset - offsetHeading),
-                                      offsetHeading);
-
-      m_CorrectedPose = rScanPose - worldSensorOffset;
+      m_CorrectedPose = GetCorrectedAt(rScanPose);
 
       Update();
     }
@@ -5656,6 +5647,25 @@ namespace karto
     inline Pose2 GetSensorAt(const Pose2& rPose) const
     {
       return Transform(rPose).TransformPose(GetLaserRangeFinder()->GetOffsetPose());
+    }
+
+    /**
+     * @brief Computes the pose of the robot if the sensor were at the given pose
+     * @param sPose sensor pose
+     * @return robot pose
+     */
+    inline Pose2 GetCorrectedAt(const Pose2& sPose) const
+    {
+      Pose2 deviceOffsetPose2 = GetLaserRangeFinder()->GetOffsetPose();
+      kt_double offsetLength = deviceOffsetPose2.GetPosition().Length();
+      kt_double offsetHeading = deviceOffsetPose2.GetHeading();
+      kt_double angleoffset = atan2(deviceOffsetPose2.GetY(), deviceOffsetPose2.GetX());
+      kt_double correctedHeading = math::NormalizeAngle(sPose.GetHeading());
+      Pose2 worldSensorOffset = Pose2(offsetLength * cos(correctedHeading + angleoffset - offsetHeading),
+                                      offsetLength * sin(correctedHeading + angleoffset - offsetHeading),
+                                      offsetHeading);
+
+      return sPose - worldSensorOffset;
     }
 
     /**

--- a/slam_toolbox/lib/karto_sdk/include/karto_sdk/Karto.h
+++ b/slam_toolbox/lib/karto_sdk/include/karto_sdk/Karto.h
@@ -5572,6 +5572,17 @@ namespace karto
 
       m_IsDirty = true;
     }
+    
+    /**
+     * Moves the scan by moving the robot pose to the given location and update point readings.
+     * @param rPose new pose of the robot of this scan
+     */
+    inline void SetCorrectedPoseAndUpdate(const Pose2& rPose)
+    {
+      SetCorrectedPose(rPose);
+      
+      Update();
+    }
 
     /**
      * Gets barycenter of point readings

--- a/slam_toolbox/lib/karto_sdk/src/Mapper.cpp
+++ b/slam_toolbox/lib/karto_sdk/src/Mapper.cpp
@@ -1605,7 +1605,11 @@ namespace karto
     // only attach link information if the edge is new
     if (isNewEdge == true)
     {
-      pEdge->SetLabel(new LinkInfo(pFromScan->GetSensorPose(), rMean, rCovariance));
+      auto corrected_pose=pToScan->GetCorrectedPose();
+      pToScan->SetSensorPose(rMean);
+      auto corrected_mean_pose=pToScan->GetCorrectedPose();
+      pToScan->SetCorrectedPose(corrected_pose);
+      pEdge->SetLabel(new LinkInfo(pFromScan->GetCorrectedPose(), corrected_mean_pose, rCovariance));
       if (m_pMapper->m_pScanOptimizer != NULL)
       {
         m_pMapper->m_pScanOptimizer->AddConstraint(pEdge);
@@ -1988,7 +1992,7 @@ namespace karto
         {
           continue;
         }
-        scan->SetSensorPose(iter->second);
+        scan->SetCorrectedPose(iter->second);
       }
 
       pSolver->Clear();

--- a/slam_toolbox/lib/karto_sdk/src/Mapper.cpp
+++ b/slam_toolbox/lib/karto_sdk/src/Mapper.cpp
@@ -1988,7 +1988,7 @@ namespace karto
         {
           continue;
         }
-        scan->SetCorrectedPose(iter->second);
+        scan->SetCorrectedPoseAndUpdate(iter->second);
       }
 
       pSolver->Clear();

--- a/slam_toolbox/lib/karto_sdk/src/Mapper.cpp
+++ b/slam_toolbox/lib/karto_sdk/src/Mapper.cpp
@@ -1605,11 +1605,7 @@ namespace karto
     // only attach link information if the edge is new
     if (isNewEdge == true)
     {
-      Pose2 corrected_pose = pToScan->GetCorrectedPose();
-      pToScan->SetSensorPose(rMean);
-      Pose2 corrected_mean_pose = pToScan->GetCorrectedPose();
-      pToScan->SetCorrectedPose(corrected_pose);
-      pEdge->SetLabel(new LinkInfo(pFromScan->GetCorrectedPose(), corrected_mean_pose, rCovariance));
+      pEdge->SetLabel(new LinkInfo(pFromScan->GetCorrectedPose(), pToScan->GetCorrectedAt(rMean), rCovariance));
       if (m_pMapper->m_pScanOptimizer != NULL)
       {
         m_pMapper->m_pScanOptimizer->AddConstraint(pEdge);

--- a/slam_toolbox/lib/karto_sdk/src/Mapper.cpp
+++ b/slam_toolbox/lib/karto_sdk/src/Mapper.cpp
@@ -1605,9 +1605,9 @@ namespace karto
     // only attach link information if the edge is new
     if (isNewEdge == true)
     {
-      auto corrected_pose=pToScan->GetCorrectedPose();
+      Pose2 corrected_pose = pToScan->GetCorrectedPose();
       pToScan->SetSensorPose(rMean);
-      auto corrected_mean_pose=pToScan->GetCorrectedPose();
+      Pose2 corrected_mean_pose = pToScan->GetCorrectedPose();
       pToScan->SetCorrectedPose(corrected_pose);
       pEdge->SetLabel(new LinkInfo(pFromScan->GetCorrectedPose(), corrected_mean_pose, rCovariance));
       if (m_pMapper->m_pScanOptimizer != NULL)

